include/openthread/tcp.h

```diff
@@ -0,0 +1,589 @@
+ *
+ * @retval OT_ERROR_NONE    Successfully added data to the send buffer.
+ * @retval OT_ERROR_FAILED  Failed to add data to the send buffer.
+ */
+otError otTcpActiveSendByReference(otInstance *aInstance, otTcpActiveSocket *aSocket, otLinkedBuffer *aBuffer, uint32_t aFlags);
+
+/**
+ * Adds data to the send buffer by extending the length of the final
+ * otLinkedBuffer in the send buffer by the specified amount. If the send
+ * buffer is empty, then the operation fails.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket    A pointer to the active TCP socket structure representing the TCP endpoint on which to send data.
+ * @param[in]  aNumBytes  The number of bytes by which to extend the length of the final linked buffer.
+ * @param[in]  aFlags     Flags specifying options for this operation (see enumeration above).
+ *
+ * @retval OT_ERROR_NONE    Successfully added data to the send buffer.
+ * @retval OT_ERROR_FAILED  Failed to add data to the send buffer.
+ */
+otError otTcpActiveSendByExtension(otInstance *aInstance, otTcpActiveSocket *aSocket, size_t aNumBytes, uint32_t aFlags);
+
+/**
+ * Provides the application with a linked buffer chain referencing data
+ * currently in the TCP receive buffer. The linked buffer chain is valid until
+ * the "receive ready" callback is next invoked, or until the next call to
+ * @p otTcpActiveReceiveContiguify or @p otTcpActiveCommitReceive.
+ *
+ * @param[in]   aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]   aSocket    A pointer to the active TCP socket structure representing the TCP endpoint on which to receive data.
+ * @param[out]  aBuffer    A pointer to the linked buffer chain referencing data currently in the receive buffer.
+ *
+ * @retval OT_ERROR_NONE    Successfully completed the operation.
+ * @retval OT_ERROR_FAILED  Failed to complete the operation.
+ */
+otError otTcpActiveReceiveByReference(otInstance *aInstance, otTcpActiveSocket *aSocket, const otLinkedBuffer **aBuffer);
+
+/**
+ * Reorganizes the receive buffer to be entirely contiguous in memory
+ *
+ * This is optional; an application can simply traverse the linked buffer
+ * chain obtained by calling @p otTcpActiveReceiveByReference. Some
+ * applications may wish to call this function to make the receive buffer
+ * contiguous to simplify their data processing, but this comes at the expense
+ * of CPU time to reorganize the data in the receive buffer.
+ *
+ * @param[in]   aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]   aSocket    A pointer to the active TCP socket structure representing the TCP endpoint on which to receive data.
+ *
+ * @retval OT_ERROR_NONE    Successfully completed the operation.
+ * @retval OT_ERROR_FAILED  Failed to complete the operation.
+ */
+otError otTcpActiveReceiveContiguify(otInstance *aInstance, otTcpActiveSocket *aSocket);
```

Comment:
Is ALP guaranteed to get one single `otLinkedBuffer` after this call by `otTcpActiveReceiveByReference` ?

Reply:
Yes, that's correct.